Error running default wry template

**Describe the bug**
When I generate a wry template using the latest cargo-mobile2 (or version 0.5.1 as suggested here : https://github.com/tauri-apps/cargo-mobile2/issues/210#issuecomment-1728503050), I get this error :
e: /home/riccardo/dioxus-mobile-test/gen/android/app/src/main/kotlin/com/example/dioxus_mobile_test/MainActivity.kt: (3, 22): Unresolved reference: WryActivity

**Steps To Reproduce**
- Setup Android SDK, NDK...
- Add Rust targets
- Create a new Rust project and `cd` into it
- Run `cargo install --git https://github.com/tauri-apps/cargo-mobile2` (or `cargo install --git https://github.com/tauri-apps/cargo-mobile2 --tag tauri-mobile-v0.5.1`)
- Run `cargo mobile init`
- Run `cargo android run`

**Expected behavior**
The code to compile and run.

~~**Screenshots**~~
~~If applicable, add screenshots to help explain your problem.~~

**Platform and Versions (please complete the following information):**
Host OS: ArchLinux x86_64
Target OS: aarch64-linux-android
Rustc: rustc 1.74.1 (a28077b28 2023-12-04)
Ouput of `cargo mobile doctor`:                <!-- (if you build with specific commit, please list it as well) -->
For version 0.5.1 :
```
[!] cargo-mobile v0.5.1
    • Contains commits up to "Publish New Versions (#158)\n"
    • Installed at "~/.cargo/.tauri-mobile"
    ✗ Failed to get OS info: Failed to find `VERSION` in
      "/etc/os-release": "NAME=\"Arch Linux\"\nPRETTY_NAME=\"Arch
      Linux\"\nID=arch\nBUILD_ID=rolling\nANSI_COLOR=\"38;2;23;147;209\"
      \nHOME_URL=\"https://archlinux.org/\"\nDOCUMENTATION_URL=\"https:/
      /wiki.archlinux.org/\"\nSUPPORT_URL=\"https://bbs.archlinux.org/\"
      \nBUG_REPORT_URL=\"https://bugs.archlinux.org/\"\nPRIVACY_POLICY_U
      RL=\"https://terms.archlinux.org/docs/privacy-policy/\"\nLOGO=arch
      linux-logo\n"
    • rustc v1.74.1 (a28077b28 2023-12-4)

[✔] Android developer tools
    • SDK v26.1.1 installed at "~/Android/Sdk"
    • NDK v26.1.10909125 installed at "~/Android/Sdk/ndk/26.1.10909125"

[✔] Connected devices
    • Redmi Note 8
```
For version 0.9.1 :
```

[!] cargo-mobile v0.9.1
    • Contains commits up to "Publish New Versions (#266)\n"
    • Installed at "~/.cargo/.cargo-mobile2"
    ✗ Failed to get OS info: Failed to find `VERSION` in "/etc/os-release": "NAME=\"Arch Linux\"\nPRETTY_NAME=\"Arch
      Linux\"\nID=arch\nBUILD_ID=rolling\nANSI_COLOR=\"38;2;23;147;209\"\nHOME_URL=\"https://archlinux.org/\"\nDOCUMENTATION_URL=\"https://
      wiki.archlinux.org/\"\nSUPPORT_URL=\"https://bbs.archlinux.org/\"\nBUG_REPORT_URL=\"https://bugs.archlinux.org/
      \"\nPRIVACY_POLICY_URL=\"https://terms.archlinux.org/docs/privacy-policy/\"\nLOGO=archlinux-logo\n"
    • rustc v1.74.1 (a28077b28 2023-12-4)

[✔] Android developer tools
    • SDK v26.1.1 installed at "~/Android/Sdk"
    • NDK v26.1.10909125 installed at "~/Android/Sdk/ndk/26.1.10909125"

[✔] Connected devices
    • Redmi Note 8
```


~~**Additional context**~~
~~Add any other context about the problem here.~~

:thinking: 
I ran "cargo android apk build" and then "cargo android run" and it worked :thinking: 
I think I also reinstalled the cargo-mobile2 CLI just after `cargo mobile init` :thinking: 
I got an error few seconds after launching but it works :thinking:
```
Starting: Intent { cmp=com.example.truc/.MainActivity }
java.lang.SecurityException: Injecting to another application requires INJECT_EVENTS permission
	at android.os.Parcel.createExceptionOrNull(Parcel.java:2376)
	at android.os.Parcel.createException(Parcel.java:2360)
	at android.os.Parcel.readException(Parcel.java:2343)
	at android.os.Parcel.readException(Parcel.java:2285)
	at android.hardware.input.IInputManager$Stub$Proxy.injectInputEvent(IInputManager.java:1012)
	at android.hardware.input.InputManager.injectInputEvent(InputManager.java:907)
	at com.android.commands.input.Input.injectKeyEvent(Input.java:355)
	at com.android.commands.input.Input.access$100(Input.java:41)
	at com.android.commands.input.Input$InputKeyEvent.sendKeyEvent(Input.java:203)
	at com.android.commands.input.Input$InputKeyEvent.run(Input.java:190)
	at com.android.commands.input.Input.onRun(Input.java:107)
	at com.android.internal.os.BaseCommand.run(BaseCommand.java:60)
	at com.android.commands.input.Input.main(Input.java:71)
	at com.android.internal.os.RuntimeInit.nativeFinishInit(Native Method)
	at com.android.internal.os.RuntimeInit.main(RuntimeInit.java:463)
Caused by: android.os.RemoteException: Remote stack trace:
	at com.android.server.input.InputManagerService.injectInputEventInternal(InputManagerService.java:790)
	at com.android.server.input.InputManagerService.injectInputEvent(InputManagerService.java:764)
	at android.hardware.input.IInputManager$Stub.onTransact(IInputManager.java:459)
	at android.os.Binder.execTransactInternal(Binder.java:1154)
	at android.os.Binder.execTransact(Binder.java:1123)

--------- beginning of main
--------- beginning of system
logcat: Unexpected EOF!

This means that either the device shut down, logd crashed, or this instance of logcat was unable to read log
messages as quickly as they were being produced.

If you have enabled significant logging, look into using the -G option to increase log buffer sizes.
error: IO error
    command ["/home/riccardo/Android/Sdk/platform-tools/adb", "logcat",
    "-v", "color", "-s", "truc:W"] exited with code 1
```

With this lib.rs, it compiles, but it crashes :
```rust
use anyhow::Result;

#[cfg(target_os = "android")]
fn init_logging() {
    android_logger::init_once(
        android_logger::Config::default()
            .with_min_level(log::Level::Trace)
            .with_tag("dioxus-mobile-test"), // This name may be different for your application
    );
}

#[cfg(not(target_os = "android"))]
fn init_logging() {
    env_logger::init();
}

#[cfg(any(target_os = "android", target_os = "ios"))]
fn stop_unwind<F: FnOnce() -> T, T>(f: F) -> T {
    match std::panic::catch_unwind(std::panic::AssertUnwindSafe(f)) {
        Ok(t) => t,
        Err(err) => {
            eprintln!("attempt to unwind out of `rust` with err: {:?}", err);
            std::process::abort()
        }
    }
}

#[cfg(any(target_os = "android", target_os = "ios"))]
fn _start_app() {
    stop_unwind(|| main().unwrap());
}

#[no_mangle]
#[inline(never)]
#[cfg(target_os = "android")]
pub extern "C" fn start_app() {
    #[cfg(target_os = "android")]
    //android_binding!(com_example, dioxus_mobile_test, _start_app); // This code may be different for your application
    tao::android_binding!(
        com_example,
        wry_app,
        WryActivity,
        wry::android_setup, // pass the wry::android_setup function to tao which will invoke when the event loop is created
        _start_app
    );
    wry::android_binding!(com_example, ttt);
}

use dioxus::prelude::*;

pub fn main() -> Result<()> {
    // Right now we're going through dioxus-desktop but we'd like to go through dioxus-mobile
    // That will seed the index.html with some fixes that prevent the page from scrolling/zooming etc
    dioxus_desktop::launch::launch(
        app,
        Vec::new(),
        // Note that we have to disable the viewport goofiness of the browser.
        // Dioxus_mobile should do this for us
        dioxus_desktop::Config::default().with_custom_index(r#"<!DOCTYPE html>
        <html>
          <head>
            <title>Dioxus app</title>
            <meta name="viewport" content="width=device-width, initial-scale=1.0, maximum-scale=1.0, user-scalable=no" />
            <!-- CUSTOM HEAD -->
          </head>
          <body>
            <div id="main"></div>
            <!-- MODULE LOADER -->
          </body>
        </html>
       "#.into()),
    );

    Ok(())
}

fn app() -> Element {
    //let items = use_ref(cx, || vec![1, 2, 3]);

    log::debug!("Hello from the app");

    rsx! {
        div {
            h1 { "Hello, Mobile"}
            div { margin_left: "auto", margin_right: "auto", width: "200px", padding: "10px", border: "1px solid black",
                button {
                    onclick: move|_| {
                        println!("Clicked!");
                        // let mut items_mut = items.write();
                        // let new_item = items_mut.len() + 1;
                        // items_mut.push(new_item);
                        // println!("Requested update");
                    },
                    "Add item"
                }
                // for item in items.read().iter() {
                //     div { "- {item}" }
                // }
            }
        }
    }
}
```

```
java.lang.SecurityException: Injecting to another application requires INJECT_EVENTS permission
        at android.os.Parcel.createExceptionOrNull(Parcel.java:2376)
        at android.os.Parcel.createException(Parcel.java:2360)
        at android.os.Parcel.readException(Parcel.java:2343)
        at android.os.Parcel.readException(Parcel.java:2285)
        at android.hardware.input.IInputManager$Stub$Proxy.injectInputEvent(IInputManager.java:1012)
        at android.hardware.input.InputManager.injectInputEvent(InputManager.java:907)
        at com.android.commands.input.Input.injectKeyEvent(Input.java:355)
        at com.android.commands.input.Input.access$100(Input.java:41)
        at com.android.commands.input.Input$InputKeyEvent.sendKeyEvent(Input.java:203)
        at com.android.commands.input.Input$InputKeyEvent.run(Input.java:190)
        at com.android.commands.input.Input.onRun(Input.java:107)
        at com.android.internal.os.BaseCommand.run(BaseCommand.java:60)
        at com.android.commands.input.Input.main(Input.java:71)
        at com.android.internal.os.RuntimeInit.nativeFinishInit(Native Method)
        at com.android.internal.os.RuntimeInit.main(RuntimeInit.java:463)
Caused by: android.os.RemoteException: Remote stack trace:
        at com.android.server.input.InputManagerService.injectInputEventInternal(InputManagerService.java:790)
        at com.android.server.input.InputManagerService.injectInputEvent(InputManagerService.java:764)
        at android.hardware.input.IInputManager$Stub.onTransact(IInputManager.java:459)
        at android.os.Binder.execTransactInternal(Binder.java:1154)
        at android.os.Binder.execTransact(Binder.java:1123)
```

--- a/.changes/wry-0.37.md
+++ b/.changes/wry-0.37.md
@@ -1,0 +1,5 @@
+---
+"cargo-mobile2": "patch"
+---
+
+Update `wry` template to `wry@0.37`

--- a/templates/apps/wry/Cargo.toml.hbs
+++ b/templates/apps/wry/Cargo.toml.hbs
@@ -33,13 +33,14 @@ WRY_ANDROID_KOTLIN_FILES_OUT_DIR = "<android-project-dir>/app/src/main/kotlin/{{
 frameworks = [ "WebKit" ]
 
 [dependencies]
-anyhow = "1.0.56"
-log = "0.4.11"
-wry = "0.28.0"
+anyhow = "1.0"
+log = "0.4"
+wry = "0.37"
+tao = "0.26"
 
 [target.'cfg(target_os = "android")'.dependencies]
-android_logger = "0.9.0"
-jni = "0.19.0"
+android_logger = "0.13"
+jni = "0.21.0"
 paste = "1.0"
 
 [target.'cfg(not(target_os = "android"))'.dependencies]

--- a/templates/apps/wry/gen/bin/desktop.rs.hbs
+++ b/templates/apps/wry/gen/bin/desktop.rs.hbs
@@ -1,5 +1,5 @@
 
 fn main() {
     #[cfg(not(any(target_os = "android", target_os = "ios")))]
-    {{snake-case app.name}}::main().unwrap();
+    {{snake-case app.name}}::main();
 }

--- a/templates/apps/wry/src/lib.rs.hbs
+++ b/templates/apps/wry/src/lib.rs.hbs
@@ -1,21 +1,15 @@
-use anyhow::Result;
-#[cfg(target_os = "android")]
-use wry::android_binding;
-use wry::{
-    application::{
-        event::{Event, StartCause, WindowEvent},
-        event_loop::{ControlFlow, EventLoop, EventLoopWindowTarget},
-        window::WindowBuilder,
-    },
-    http::Response,
-    webview::{WebView, WebViewBuilder},
+use tao::{
+    event::{Event, StartCause, WindowEvent},
+    event_loop::{ControlFlow, EventLoop, EventLoopWindowTarget},
+    window::WindowBuilder,
 };
+use wry::{http, WebView, WebViewBuilder};
 
 #[cfg(target_os = "android")]
 fn init_logging() {
     android_logger::init_once(
         android_logger::Config::default()
-            .with_min_level(log::Level::Trace)
+            .with_max_level(log::LevelFilter::Trace)
             .with_tag("{{app.name}}"),
     );
 }
@@ -38,7 +32,7 @@ fn stop_unwind<F: FnOnce() -> T, T>(f: F) -> T {
 
 #[cfg(any(target_os = "android", target_os = "ios"))]
 fn _start_app() {
-    stop_unwind(|| main().unwrap());
+    stop_unwind(|| main());
 }
 
 #[no_mangle]
@@ -46,12 +40,22 @@ fn _start_app() {
 #[cfg(any(target_os = "android", target_os = "ios"))]
 pub extern "C" fn start_app() {
     #[cfg(target_os = "android")]
-    android_binding!({{reverse-domain-snake-case app.domain}}, {{snake-case app.name}}, _start_app);
+    {
+        tao::android_binding!(
+            {{reverse-domain-snake-case app.domain}},
+            {{snake-case app.name}},
+            WryActivity,
+            wry::android_setup, // pass the wry::android_setup function to tao which will invoke when the event loop is created
+            _start_app
+        );
+        wry::android_binding!(com_example, {{snake-case app.name}});
+    }
+
     #[cfg(target_os = "ios")]
     _start_app()
 }
 
-pub fn main() -> Result<()> {
+pub fn main() {
     init_logging();
     let event_loop = EventLoop::new();
 
@@ -75,64 +79,78 @@ pub fn main() -> Result<()> {
     });
 }
 
-fn build_webview(event_loop: &EventLoopWindowTarget<()>) -> Result<WebView> {
+fn build_webview(event_loop: &EventLoopWindowTarget<()>) -> anyhow::Result<WebView> {
     let window = WindowBuilder::new()
         .with_title("A fantastic window!")
         .build(&event_loop)?;
-    let webview = WebViewBuilder::new(window)?
-        .with_url("https://tauri.app")?
+    let webview = WebViewBuilder::new(&window)
+        .with_url("https://tauri.app")
         // If you want to use custom protocol, set url like this and add files like index.html to assets directory.
         // .with_url("wry://assets/index.html")?
         .with_devtools(true)
         .with_initialization_script("console.log('hello world from init script');")
-        .with_ipc_handler(|_, s| {
+        .with_ipc_handler(|s| {
             dbg!(s);
         })
-        .with_custom_protocol("wry".into(), move |_request| {
-            #[cfg(not(target_os = "android"))]
-            {
-                use std::fs::{canonicalize, read};
-                use wry::http::header::CONTENT_TYPE;
-
-                // Remove url scheme
-                let path = _request.uri().path();
-
-                #[cfg(not(target_os = "ios"))]
-                let content = read(canonicalize(&path[1..])?)?;
-
-                #[cfg(target_os = "ios")]
-                let content = {
-                    let path = core_foundation::bundle::CFBundle::main_bundle()
-                        .resources_path()
-                        .unwrap()
-                        .join(&path);
-                    read(canonicalize(&path)?)?
-                };
-
-                // Return asset contents and mime types based on file extentions
-                // If you don't want to do this manually, there are some crates for you.
-                // Such as `infer` and `mime_guess`.
-                let (data, meta) = if path.ends_with(".html") {
-                    (content, "text/html")
-                } else if path.ends_with(".js") {
-                    (content, "text/javascript")
-                } else if path.ends_with(".png") {
-                    (content, "image/png")
-                } else {
-                    unimplemented!();
-                };
-
-                Ok(Response::builder()
-                    .header(CONTENT_TYPE, meta)
-                    .body(data.into())?)
-            }
-
-            #[cfg(target_os = "android")]
-            {
-                Ok(Response::builder().body(Vec::new().into())?)
+        .with_custom_protocol("wry".into(), move |request| {
+            match process_custom_protcol(request) {
+                Ok(r) => r.map(Into::into),
+                Err(e) => http::Response::builder()
+                    .header(http::header::CONTENT_TYPE, "text/plain")
+                    .status(500)
+                    .body(e.to_string().as_bytes().to_vec())
+                    .unwrap()
+                    .map(Into::into),
             }
         })
         .build()?;
 
-        Ok(webview)
- }
+    Ok(webview)
+}
+
+fn process_custom_protcol(
+    _request: http::Request<Vec<u8>>,
+) -> anyhow::Result<http::Response<Vec<u8>>> {
+    #[cfg(not(target_os = "android"))]
+    {
+        use std::fs::{canonicalize, read};
+        use wry::http::header::CONTENT_TYPE;
+
+        // Remove url scheme
+        let path = _request.uri().path();
+
+        #[cfg(not(target_os = "ios"))]
+        let content = read(canonicalize(&path[1..])?)?;
+
+        #[cfg(target_os = "ios")]
+        let content = {
+            let path = core_foundation::bundle::CFBundle::main_bundle()
+                .resources_path()
+                .unwrap()
+                .join(&path);
+            read(canonicalize(&path)?)?
+        };
+
+        // Return asset contents and mime types based on file extentions
+        // If you don't want to do this manually, there are some crates for you.
+        // Such as `infer` and `mime_guess`.
+        let (data, meta) = if path.ends_with(".html") {
+            (content, "text/html")
+        } else if path.ends_with(".js") {
+            (content, "text/javascript")
+        } else if path.ends_with(".png") {
+            (content, "image/png")
+        } else {
+            unimplemented!();
+        };
+
+        Ok(http::Response::builder()
+            .header(CONTENT_TYPE, meta)
+            .body(data.into())?)
+    }
+
+    #[cfg(target_os = "android")]
+    {
+        Ok(http::Response::builder().body(Vec::new().into())?)
+    }
+}

